### PR TITLE
Sockops 1

### PR DESCRIPTION
**Pull request by shankarseal**
## Description

First part of #786 . Adds boiler plate code for the new program and attach types.

## Testing

A sample eBPF program is added and unit test variations added that loads the program.

## Documentation

Doxygen.

### CODE REVIEW

File: include/ebpf_nethooks.h [hunk 1]
```diff
@@ -149,9 +149,6 @@ typedef struct bpf_sock_addr
     uint32_t protocol;       ///< IP protocol.
     uint32_t compartment_id; ///< Network compartment Id.
 } bpf_sock_addr_t;
-#ifdef _MSC_VER
-#pragma warning(pop)
```
Comment (Alan-Jowett) [inline include/ebpf_nethooks.h:153]: Why was this removed?

File: include/bpf_helpers_platform.h [hunk 2]
```diff
@@ -17,4 +17,4 @@ typedef uint16_t __u16;
 typedef uint32_t __be32;
 typedef uint32_t __u32;
 typedef uint32_t __wsum;
-typedef uint64_t __u64;
+typedef uint64_t __u64;
```
Comment (dthaler) [inline include/bpf_helpers_platform.h:20]: Revert?

File: include/ebpf_nethooks.h [hunk 3]
```diff
@@ -172,3 +169,61 @@ typedef struct bpf_sock_addr
  */
 typedef int
 sock_addr_hook_t(bpf_sock_addr_t* context);
+
+typedef enum _bpf_sock_op_type
+{
+    BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB,  ///< Calls BPF program when an
+                                         ///< active (outbound) connection is
+                                         ///< established
+                                         ///<
```
Comment (dthaler) [inline include/ebpf_nethooks.h:178]: the `///<` format won't work, move any multi-line comments to lines before and use `/** @brief multi line comment here */` like I did in PR #940

File: include/ebpf_structs.h [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 #if !defined(NO_CRT)
 #include <stdint.h>
+#include <stdbool.h>
```
Comment (dthaler) [inline include/ebpf_structs.h:11]: nit: move up a line to keep alphabetical order

File: libs/api_common/windows_platform_common.cpp [hunk 5]
```diff
@@ -79,6 +79,19 @@ const EbpfProgramType windows_sock_addr_program_type = {
...
     {"cgroup/connect4", "cgroup/connect6", "cgroup/recv_accept4", "cgroup/recv_accept6"}};
 
+//
+// SOCK_OPS.
+//
+const ebpf_context_descriptor_t g_sock_ops_context_descriptor = {
+    sizeof(bpf_sock_ops_t),
+    -1, // Offset into ctx struct for pointer to data, or -1 if none.
+    -1, // Offset into ctx struct for pointer to data, or -1 if none.
+    -1, // Offset into ctx struct for pointer to metadata, or -1 if none.
```
Comment (dthaler) [inline libs/api_common/windows_platform_common.cpp:89]: this is also duplicated in net_ebpf_ext_program_info.h.  Is it easy to have it be in one place instead of duplicated in both places?  If not easy, then ok as is.

File: include/ebpf_nethooks.h (hunk 1 above)
Comment (shankarseal) [reply include/ebpf_nethooks.h:153]: moved further below, not removed.

File: libs/api_common/windows_platform_common.cpp (hunk 5 above)
Comment (shankarseal) [reply libs/api_common/windows_platform_common.cpp:89]: There was a similar feedback for SOCK_ADDR program context descriptor as well. I think having a common header was problematic. Since this should be addressed by #223, I am leaving this as is.

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]
